src/gc.c

```diff
@@ -539,16 +540,28 @@ static void run_finalizers(void)
     arraylist_free(&copied_list);
 }
 
-void jl_gc_inhibit_finalizers(int state)
+JL_DLLEXPORT int jl_finalizers_enable(int on)
 {
-    // NOTE: currently only called with the codegen lock held, but might need
-    // more synchronization in the future
-    if (jl_gc_finalizers_inhibited && !state && !jl_in_finalizer) {
-        jl_in_finalizer = 1;
-        run_finalizers();
-        jl_in_finalizer = 0;
+    jl_tls_states_t *ptls = jl_get_ptls_states();
+    int prev = !ptls->disable_finalizers;
+    ptls->disable_finalizers = (on == 0);
+    if (on && !prev) {
+        // disable -> enable
+        // TODO: the atomics are in the wrong place / missing some thread locks
+        jl_atomic_fetch_add(&jl_finalizers_disabled_counter, -1);
+        if (jl_finalizers_disabled_counter == 0) {
+            int8_t was_in_finalizer = jl_in_finalizer;
+            jl_in_finalizer = 1;
+            run_finalizers();
+            jl_in_finalizer = was_in_finalizer;
+        }
     }
-    jl_gc_finalizers_inhibited = state;
+    else if (prev && !on) {
+        // enable -> disable
+        jl_atomic_fetch_add(&jl_finalizers_disabled_counter, 1);
+        // TODO: check if finalizers are running and wait for them to finish
```

Comment:
Is this really necessary? Since finalizer runs can nest and can run on multiple threads, I feel like waiting for all finalizers to finish is asking for dead lock. (e.g. if this is called by a finalizer on two different threads because both of them modifies a dict).

For avoiding corruption with multithreading it is still necessary to use explicit locks. It is only the case if the finalizer runs on the same thread that you can't really protect with a lock and requires this API to disable the finalizer on the same thread.